From eaef9b4cdcc2128518902fe37910a2e6b1e9e0ec Mon Sep 17 00:00:00 2001
From: Yan Markman <ymarkman@marvell.com>
Date: Tue, 12 Jul 2016 18:13:40 +0300
Subject: [PATCH 0586/2241] pp3: use sw-checksum for tcp over nss

In path Host->nss->wlan the software IPv4/v6 UDP/TCP checksum
should be used instead of hw-checksum
Do not set (NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM) for "nss"
in procedure mv_pp3_late_init().

Change-Id: Ie01e600454e7c91a62613bb487592afbd8e18062
Signed-off-by: Yan Markman <ymarkman@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31097
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Dovrat Zifroni <dovrat@marvell.com>
Tested-by: Dovrat Zifroni <dovrat@marvell.com>
Reviewed-by: Yelena Krivosheev <yelena@marvell.com>
---
 drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
index b0dbb819..5b7d8d0 100644
--- a/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
+++ b/drivers/net/ethernet/marvell/pp3/net_dev/mv_netdev.c
@@ -2097,9 +2097,16 @@ int mv_pp3_cpu_affinity_set(struct net_device *dev, int cpu)
  *----------------------------------------------------------------------------*/
 static int mv_pp3_late_init(struct net_device *dev)
 {
-	dev->features = NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM | NETIF_F_RXCSUM;
-
-	dev->hw_features |= NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM | NETIF_F_RXCSUM;
+	dev->features = NETIF_F_RXCSUM;
+	dev->hw_features |= NETIF_F_RXCSUM;
+	/* Add HW-checksum for NOT "nssN" devices only.
+	 * NSS type is dev_priv->vport->type=MV_PP3_NSS_PORT_EXT
+	 * but dev_priv->vport==NULL here.
+	 */
+	if (!strstr(dev->name, "nss")) {
+		dev->features |= NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM;
+		dev->hw_features |= NETIF_F_IP_CSUM | NETIF_F_IPV6_CSUM;
+	}
 
 #ifdef CONFIG_MV_PP3_SG
 	dev->features |= NETIF_F_SG;
-- 
2.7.4

